ssh/tailssh/tailssh.go

```diff
@@ -288,6 +290,10 @@ func (c *conn) clientAuth(cm gossh.ConnMetadata) (*gossh.Permissions, error) {
 		return &gossh.Permissions{}, nil
 	}
 
+	if !c.clientAuthAttempted.CompareAndSwap(false, true) {
+		panic("clientAuth unexpectedly called twice")
```

Comment:
can't they call NoClientAuthCallback and then later password or PublicKeyCallback?

Reply:
when any of the callbacks returns a `nil` error, `ssh.Server` will stop the auth loop and shouldn't call any other callbacks: https://github.com/golang/crypto/blob/master/ssh/server.go#L798-L800

Reply:
oh, and if they hit the `+password` hack, we swap the password callback for this one: https://github.com/tailscale/tailscale/blob/dd7ce90987369a1c146dee8308ca4b34c2487a66/ssh/tailssh/tailssh.go#L387-L395

Reply:
See [here](https://cs.opensource.google/go/x/crypto/+/refs/tags/v0.33.0:ssh/server.go;l=540) for the controlling loop. By default, the SSH server will try authentication up to 6 times. Each time, it reads an [auth request](https://cs.opensource.google/go/x/crypto/+/refs/tags/v0.33.0:ssh/server.go;l=556) from the client and then tries to do the authentication. If it succeeds, it [breaks out of the loop](https://cs.opensource.google/go/x/crypto/+/refs/tags/v0.33.0:ssh/server.go;l=799), and we won't get any subsequent authentication attempts.

Reply:
This behavior seems to be spec compliant. RFC4252 Section 5 says

```
While there is usually little point for clients to send requests that
the server does not list as acceptable, sending such requests is not
an error, and the server SHOULD simply reject requests that it does
not recognize.
```

Also

```
the client MAY at any time continue with a new
SSH_MSG_USERAUTH_REQUEST message, in which case the server MUST
abandon the previous authentication attempt and continue with the new
one.
```